crates/ruff/src/rules/flake8_todos/rules/todos.rs

```diff
@@ -296,25 +239,27 @@ static ISSUE_LINK_REGEX_SET: Lazy<RegexSet> = Lazy::new(|| {
     .unwrap()
 });
 
-pub(crate) fn todos(indexer: &Indexer, locator: &Locator, settings: &Settings) -> Vec<Diagnostic> {
+pub(crate) fn todos(
+    comment_directive_ranges: Vec<(TextRange, usize, TextRange)>,
+    indexer: &Indexer,
+    locator: &Locator,
+    settings: &Settings,
+) -> Vec<Diagnostic> {
     let mut diagnostics: Vec<Diagnostic> = vec![];
 
-    let mut iter = indexer.comment_ranges().iter().peekable();
-    while let Some(comment_range) = iter.next() {
-        let comment = locator.slice(*comment_range);
-
-        // Check that the comment is a TODO (properly formed or not).
-        let Some(tag) = detect_tag(comment, comment_range.start()) else {
-            continue;
+    for (comment_range, comment_index, directive_range) in comment_directive_ranges {
+        let comment = locator.slice(comment_range);
+        let tag = Tag {
+            content: locator.slice(directive_range),
+            range: directive_range,
         };
 
         tag_errors(&tag, &mut diagnostics, settings);
-        static_errors(&mut diagnostics, comment, *comment_range, &tag);
+        static_errors(&mut diagnostics, comment, comment_range, &tag);
 
-        // TD003
         let mut has_issue_link = false;
-        let mut curr_range = comment_range;
-        while let Some(next_range) = iter.peek() {
+        let mut curr_range = &comment_range;
+        for next_range in indexer.comment_ranges().iter().skip(comment_index + 1) {
```

Comment:
I don't think this is a problem. I just want to call it out: 
This will be slightly less efficient than the old implementation because we'll now try to extract the tag from the following comments again in the `comment_directive_ranges` loop. It also depends on `comment_ranges()` being a slice iter (or any iter with a `O(1)` skip).